src/data/list/duplicate.lean

```diff
@@ -0,0 +1,234 @@
+end
+
+lemma duplicate.of_duplicate_cons {y : α} (h : x ∈+ y :: l) (hx : x ≠ y) : x ∈+ l :=
+by simpa [duplicate_cons_iff, hx.symm] using h
+
+lemma duplicate_cons_iff_of_ne {y : α} (hne : x ≠ y) : x ∈+ y :: l ↔ x ∈+ l :=
+by simp [duplicate_cons_iff, hne.symm]
+
+lemma duplicate.mono_sublist {l' : list α} (hx : x ∈+ l) (h : l <+ l') : x ∈+ l' :=
+begin
+  induction h with l₁ l₂ y h IH l₁ l₂ y h IH,
+  { simpa using hx.ne_nil },
+  { exact (IH hx).duplicate_cons _ },
+  { rw duplicate_cons_iff at hx ⊢,
+    rcases hx with ⟨rfl, hx⟩|hx,
+    { simp [h.subset hx] },
+    { simp [IH hx] } }
+end
+
+/-- The contrapositive of `list.nodup_iff_sublist`. -/
+lemma duplicate_iff_sublist : x ∈+ l ↔ [x, x] <+ l :=
+begin
+  induction l with y l IH,
+  { simp },
+  { by_cases hx : x = y,
+    { simp [hx, cons_sublist_cons_iff, singleton_sublist] },
+    { rw [duplicate_cons_iff_of_ne hx, IH],
+      refine ⟨sublist_cons_of_sublist y, λ h, _⟩,
+      cases h,
+      { assumption },
+      { contradiction } } }
+end
+
+lemma nodup_iff_forall_not_duplicate : nodup l ↔ ∀ (x : α), ¬ x ∈+ l :=
+by simp_rw [nodup_iff_sublist, duplicate_iff_sublist]
+
+lemma duplicate.not_nodup (h : x ∈+ l) : ¬ nodup l :=
+λ H, nodup_iff_forall_not_duplicate.mp H _ h
+
+lemma duplicate_iff_count_le_two [decidable_eq α] : (x ∈+ l) ↔ 2 ≤ count x l :=
+by simp [duplicate_iff_sublist, le_count_iff_repeat_sublist]
+
+lemma sublist_of_order_embedding {l l' : list α} (f : ℕ ↪o ℕ)
+  (hf : ∀ (ix : ℕ), l.nth ix = l'.nth (f ix)) :
+  l <+ l' :=
+begin
+  induction l with hd tl IH generalizing l' f,
+  { simp },
+  have : some hd = _ := hf 0,
+  rw [eq_comm, list.nth_eq_some] at this,
+  obtain ⟨w, h⟩ := this,
+  let f' : ℕ ↪o ℕ := order_embedding.of_map_rel_iff (λ i, f (i + 1) - (f 0 + 1))
+    (λ a b, by simp [nat.sub_le_sub_right_iff, nat.succ_le_iff, nat.lt_succ_iff]),
+  have : ∀ ix, tl.nth ix = (l'.drop (f 0 + 1)).nth (f' ix),
+  { intro ix,
+    simp [list.nth_drop, nat.add_sub_of_le, nat.succ_le_iff, ←hf] },
+  rw [←list.take_append_drop (f 0 + 1) l', ←list.singleton_append],
+  apply list.sublist.append _ (IH _ this),
+  rw [list.singleton_sublist, ←h, l'.nth_le_take _ (nat.lt_succ_self _)],
+  apply list.nth_le_mem
+end
+
+lemma sublist_iff_nth_eq {l l' : list α} :
+  l <+ l' ↔ ∃ (f : ℕ ↪o ℕ), ∀ (ix : ℕ), l.nth ix = l'.nth (f ix) :=
+begin
+  split,
+  { intro H,
+    induction H with xs ys y H IH xs ys x H IH,
+    { simp },
+    { obtain ⟨f, hf⟩ := IH,
+      refine ⟨f.trans (order_embedding.of_strict_mono (+ 1) (λ _, by simp)), _⟩,
+      simpa using hf },
+    { obtain ⟨f, hf⟩ := IH,
+      refine ⟨order_embedding.of_map_rel_iff
+        (λ (ix : ℕ), if ix = 0 then 0 else (f ix.pred).succ) _, _⟩,
+      { rintro ⟨_|a⟩ ⟨_|b⟩;
+        simp [nat.succ_le_succ_iff] },
+      { rintro ⟨_|i⟩,
+        { simp },
+        { simpa using hf _ } } } },
+  { rintro ⟨f, hf⟩,
+    exact sublist_of_order_embedding f hf }
+end
+
+lemma sublist_iff_nth_le_eq {l l' : list α} :
```

Comment:
some more comments+docstrings would be nice; the statements parse but they're a lot easier to read in words